### ISSUE TITLE
collection-object-parameters-may-be-null.astub: Attempts to widen containsAll+retainAll parameter types appear not to work

### ISSUE THREAD

**Issue by cpovirk** [typetools/checker-framework#3236]
```java
$ cat ContainsAllRetainAll.java 
import java.util.Collection;

class ContainsAllRetainAll {
  void bulkOperations(Collection<?> a, Collection<?> b) {
    a.containsAll(b);
    a.retainAll(b);
  }
}
```

I would expect for this to fail with the new JDK stubs in 3.3.0 (and it does). However, I would expect it to pass when I supply `-Astubs=collection-object-parameters-may-be-null.astub`, yet it still fails:

```
$ checker/bin/javac -processor org.checkerframework.checker.nullness.NullnessChecker -Astubs=collection-object-parameters-may-be-null.astub ContainsAllRetainAll.java
warning: StubParser: annotateTypeParameters: mismatched sizes:  typeParameters (size 0)=[];  typeArguments (size 1)=[E extends Object];  decl=class AbstractCollection {      @Pure     public boolean contains(@Nullable Object o);      public boolean remove(@Nullable Object o);      @Pure     public boolean containsAll(Collection<?> c);      public boolean removeAll(Collection<?> c);      public boolean retainAll(Collection<?> c); };  elt=java.util.AbstractCollection (class com.sun.tools.javac.code.Symbol$ClassSymbol).; for more details, run with -AstubDebug
warning: StubParser: annotateTypeParameters: mismatched sizes:  typeParameters (size 0)=[];  typeArguments (size 2)=[K extends Object, V extends Object];  decl=class AbstractMap {      public boolean containsValue(@Nullable Object value);      public boolean containsKey(@Nullable Object key);      @Nullable     public V get(@Nullable Object key);      @Nullable     public V remove(@Nullable Object key); };  elt=java.util.AbstractMap (class com.sun.tools.javac.code.Symbol$ClassSymbol).; for more details, run with -AstubDebug
warning: StubParser: annotateTypeParameters: mismatched sizes:  typeParameters (size 0)=[];  typeArguments (size 1)=[E extends Object];  decl=class AbstractSet {      public boolean removeAll(Collection<?> c); };  elt=java.util.AbstractSet (class com.sun.tools.javac.code.Symbol$ClassSymbol).; for more details, run with -AstubDebug
warning: StubParser: annotateTypeParameters: mismatched sizes:  typeParameters (size 0)=[];  typeArguments (size 1)=[E extends Object];  decl=class Vector {      public synchronized boolean containsAll(Collection<?> c);      public synchronized boolean removeAll(Collection<?> c);      public synchronized boolean retainAll(Collection<?> c); };  elt=java.util.Vector (class com.sun.tools.javac.code.Symbol$ClassSymbol).; for more details, run with -AstubDebug
warning: StubParser: stub file does not match bytecode: could not find superinterface List<E> from type CopyOnWriteArraySet<E extends Object>
warning: StubParser: stub file does not match bytecode: could not find superinterface RandomAccess from type CopyOnWriteArraySet<E extends Object>
warning: StubParser: stub file does not match bytecode: could not find superinterface Cloneable from type CopyOnWriteArraySet<E extends Object>
warning: StubParser: annotateTypeParameters: mismatched sizes:  typeParameters (size 0)=[];  typeArguments (size 1)=[E extends @Initialized @Nullable Object];  decl=class AbstractCollection {      @Pure     public boolean contains(@Nullable Object o);      public boolean remove(@Nullable Object o);      @Pure     public boolean containsAll(Collection<?> c);      public boolean removeAll(Collection<?> c);      public boolean retainAll(Collection<?> c); };  elt=java.util.AbstractCollection (class com.sun.tools.javac.code.Symbol$ClassSymbol).; for more details, run with -AstubDebug
warning: StubParser: annotateTypeParameters: mismatched sizes:  typeParameters (size 0)=[];  typeArguments (size 2)=[K extends @Initialized @Nullable Object, V extends @Initialized @Nullable Object];  decl=class AbstractMap {      public boolean containsValue(@Nullable Object value);      public boolean containsKey(@Nullable Object key);      @Nullable     public V get(@Nullable Object key);      @Nullable     public V remove(@Nullable Object key); };  elt=java.util.AbstractMap (class com.sun.tools.javac.code.Symbol$ClassSymbol).; for more details, run with -AstubDebug
warning: StubParser: annotateTypeParameters: mismatched sizes:  typeParameters (size 0)=[];  typeArguments (size 1)=[E extends @Initialized @Nullable Object];  decl=class AbstractSet {      public boolean removeAll(Collection<?> c); };  elt=java.util.AbstractSet (class com.sun.tools.javac.code.Symbol$ClassSymbol).; for more details, run with -AstubDebug
warning: StubParser: annotateTypeParameters: mismatched sizes:  typeParameters (size 0)=[];  typeArguments (size 1)=[E extends @Initialized @Nullable Object];  decl=class Vector {      public synchronized boolean containsAll(Collection<?> c);      public synchronized boolean removeAll(Collection<?> c);      public synchronized boolean retainAll(Collection<?> c); };  elt=java.util.Vector (class com.sun.tools.javac.code.Symbol$ClassSymbol).; for more details, run with -AstubDebug
warning: StubParser: stub file does not match bytecode: could not find superinterface List<E> from type CopyOnWriteArraySet<E extends @Initialized @Nullable Object>
warning: StubParser: stub file does not match bytecode: could not find superinterface RandomAccess from type CopyOnWriteArraySet<E extends @Initialized @Nullable Object>
warning: StubParser: stub file does not match bytecode: could not find superinterface Cloneable from type CopyOnWriteArraySet<E extends @Initialized @Nullable Object>
ContainsAllRetainAll.java:5: error: [argument.type.incompatible] incompatible types in argument.
    a.containsAll(b);
                  ^
  found   : @Initialized @NonNull Collection<? extends @Initialized @Nullable Object>
  required: @Initialized @NonNull Collection<? extends @Initialized @NonNull Object>
ContainsAllRetainAll.java:6: error: [argument.type.incompatible] incompatible types in argument.
    a.retainAll(b);
                ^
  found   : @Initialized @NonNull Collection<? extends @Initialized @Nullable Object>
  required: @Initialized @NonNull Collection<? extends @Initialized @NonNull Object>
2 errors
14 warnings
```

I have gotten myself a little confused as I investigate: I thought I later even went and rebuilt CF's _built-in_ annotated JDK stubs with the parameter types changed to `Collection<? extends @Nullable Object>` and _still_ saw the error (with or without the `astub` file). But I wouldn't rule out the possibility that I messed something up during that. I think I at least have the original report correct :)

**Comment by cpovirk** [comment]
I am definitely confused about what workarounds I tested. I think I at least have now successfully compiled with _some_ workaround, so I will report back once I un-confuse myself :)

**Comment by cpovirk** [comment]
Correction: If I rebuild the CF's _built-in_ JDK stubs with the parameter types changed to `Collection<? extends @Nullable Object>` (or `Collection<?>`), I _can_ compile successfully. This suggests that there's a problem with using `collection-object-parameters-may-be-null.astub` (or, based on my testing, any other `astub` file) to override the CF's built-in JDK stubs. That problem is quite possibly specific to _wildcards_, as I haven't encountered problems with anything else that the stub file overrides. I will tentatively re-title this issue.

Largely for my reference: The classes I edited in the built-in stubs were:
```
AbstractCollection
AbstractSet
Collection
List
Set
```

For my test code above, I would expect for changing `Collection` alone to be sufficient.

**Comment by cpovirk** [comment]
Argh, sorry, I am wrong again:

The problem is specifically with a stub whose parameter type is `Collection<?>`. If I change the stub to have parameter type `Collection<? extends @Nullable Object>`, then I'm able to compile successfully. I thought I had tried this way back at the beginning, but I must have gotten something wrong. I will re-title the issue yet again.

This leaves us with a couple directions to go:

- Maybe the Checker Framework should treat a stub with type `Collection<?>` as overriding a source (or other stub) with type `Collection<? extends @NonNull Object>`?

- If not, or if we want a short-term fix, we could change the existing stub file to use `Collection<? extends @Nullable Object>` instead of `Collection<?>`. I'm going to make that change locally, so if you'd like a PR, let me know.

**Comment by mernst** [comment]
The problem is most likely in the Checker Framework's `StubParser.java` file.
The problem can be reproduced with a much smaller (one-line) stub file for ease of debugging.
